refactor!: use exandra

noaccOS opened the pull request:
use exandra for queries instead of `Xandra.Cluster`
make cqex an optional dependency, so that it can completely be removed by dev/release

BREAKING CHANGE: XandraUtils module has been removed
BREAKING CHANGE: `Interface.retrieve_interface_row` now returns the interface struct

depends on #92

noaccOS commented on lib/astarte_data_access/astarte/kv_store.ex:
```diff
@@ -0,0 +1,10 @@
+
+  @primary_key false
+  schema "kv_store" do
+    field(:group, :string, primary_key: true)
```
Done, I went with `:ecto` as `:exandra` didn't work

noaccOS committed:
test: add missing device attributes

Signed-off-by: Francesco Noacco <francesco.noacco@secomind.com>

noaccOS committed:
temp: chore: update astarte_core dependency

we need the updated functions which accept already casted values as well
as the database int values.

calling the functions without first updating astarte_core results in a
crash.

update this change to use mainline's astarte_core once
https://github.com/astarte-platform/astarte_core/pull/117 is merged

Signed-off-by: Francesco Noacco <francesco.noacco@secomind.com>

noaccOS committed:
feat: setup exandra

add dependencies, repo and resources

Signed-off-by: Francesco Noacco <francesco.noacco@secomind.com>

noaccOS committed:
test: wait for cluster connection on start

Signed-off-by: Francesco Noacco <francesco.noacco@secomind.com>

noaccOS committed:
refactor(device): use exandra

noaccOS committed:
refactor(interface)!: use exandra

BREAKING: Interface.retrieve_interface_row now returns the struct


Signed-off-by: Francesco Noacco <francesco.noacco@secomind.com>

noaccOS committed:
refactor(mapping): use exandra

Signed-off-by: Francesco Noacco <francesco.noacco@secomind.com>

noaccOS committed:
test: update xandra nodes settings alongside cqerl's

this also guarantees the port is always an integer

Signed-off-by: Francesco Noacco <francesco.noacco@secomind.com>

noaccOS committed:
feat(realm): expose keyspace name function

noaccOS committed:
refactor(data): use exandra

noaccOS committed:
chore: mark CASSANDRA_AUTODISCOVERY_ENABLED as deprecated

due to reworks in xandra, this option is now ignored and always enabled

Signed-off-by: Francesco Noacco <francesco.noacco@secomind.com>

Annopaolo reviewed (changes requested):
A part from some cosmetic comments, I think this change is very good and much needed. Bravo!

noaccOS commented on mix.exs:
```diff
@@ -71,7 +71,7 @@ defmodule Astarte.DataAccess.Mixfile do
   defp deps do
     [
-      {:xandra, "~> 0.11"},
+      {:exandra, github: "noaccos/exandra", ref: "push-ksolwvtqvqtr"},
```
This was probably a mistake, yes. We're able to use exandra from main here (although we need custom a exandra in https://github.com/astarte-platform/astarte/pull/1056 but that can be handled on the astarte side)

noaccOS replied:
restored mainline exandra version

noaccOS commented on lib/astarte_data_access/realms/endpoint.ex:
```diff
@@ -0,0 +1,48 @@
+  alias Astarte.Core.Mapping.Retention
+  alias Astarte.Core.Mapping.ValueType
+
+  @primary_key false
```
I agree! I was also thinking of using [TypedEctoSchema](https://hexdocs.pm/typed_ecto_schema/TypedEctoSchema.html) to do that automatically

noaccOS replied:
added typedectoschema

noaccOS committed:
chore: replace the deprecated Logger.warn with Logger.warning

Signed-off-by: Francesco Noacco <francesco.noacco@secomind.com>

noaccOS committed:
chore: fix typespec

Signed-off-by: Francesco Noacco <francesco.noacco@secomind.com>

noaccOS commented on lib/astarte_data_access/data.ex:
```diff
@@ -42,48 +63,32 @@ defmodule Astarte.DataAccess.Data do
-    case Enum.to_list(page) do
-      [] ->
-        {:error, :property_not_set}
+    data_id = data_id(device_id, interface_id, endpoint_id, path)
```
renamed as suggested

noaccOS commented on lib/astarte_data_access/mappings.ex:
```diff
@@ -17,56 +17,61 @@
+  alias Astarte.DataAccess.Astarte.Realm
+  alias Astarte.DataAccess.Realms.Endpoint
+  alias Astarte.DataAccess.Repo
   alias Astarte.DataAccess.XandraUtils
```
unused alias removed

noaccOS committed:
refactor!: remove xandra_utils

exandra provides about the same functionality as xandra_utils

furthermore, if we want to avoid having a second open connection pool to
xandra, xandra_utils would need to be refactored to use
exandra's connection pool instead

Signed-off-by: Francesco Noacco <francesco.noacco@secomind.com>

noaccOS committed:
feat: make cqex optional

allow data_access to be used without cqex/cqerl, but still allow their
usage, eg for tests

Signed-off-by: Francesco Noacco <francesco.noacco@secomind.com>
